### PR TITLE
Fix multiple definition fetch issue

### PR DESCRIPTION
**Pull request by yoching**
Hello. I updated codes to show multiple items in Workspace when the API returns multiple terms/types. This will fix  https://github.com/unisonweb/ui-core/issues/68

You can try it in the storybook with mocked API responses. However, I haven't tested it with actual UCM.

### CODE REVIEW

File: src/Code/Workspace/WorkspaceItem.elm [hunk 1]
```diff
@@ -855,6 +861,35 @@ decodeTypes ref =
     Decode.keyValuePairs decodeTypeDetails |> Decode.map buildTypes
 
 
+decodeTypesWithRef : Decode.Decoder (List ( Reference, TypeDetailWithDoc ))
+decodeTypesWithRef =
+    let
+        makeType ( hash_, d ) =
+            let
+                ref =
+                    Reference.fromFQN Reference.TypeReference d.name
```
Comment (yoching) [inline src/Code/Workspace/WorkspaceItem.elm:870]: In this part, `ref` is constructed based on `bestTypeName` (`d.name`) of the API response. 
I'm not sure if this works in all cases. Do you think this is fine?

File: src/Code/Workspace/WorkspaceItem.elm [hunk 2]
```diff
@@ -907,18 +942,63 @@ decodeTerms ref =
...
-    Decode.map2 List.append
-        (Decode.map (List.map TermItem) (field "termDefinitions" (decodeTerms ref)))
-        (Decode.map (List.map TypeItem) (field "typeDefinitions" (decodeTypes ref)))
+decodeTermsWithRef : Decode.Decoder (List ( Reference, TermDetailWithDoc ))
+decodeTermsWithRef =
+    let
+        makeTerm ( hash_, d ) =
+            let
+                ref =
+                    Reference.fromFQN Reference.TermReference d.name
```
Comment (yoching) [inline src/Code/Workspace/WorkspaceItem.elm:951]: Same here. The ref is constructed based on `bestTermName` of the API response

**Comment by yoching** [comment]
With Storybook, you can try with these buttons
Screenshot 2024-05-30 at 20 40 13

File: src/Code/Workspace/WorkspaceItem.elm (hunk 1 above)
Comment (yoching) [reply src/Code/Workspace/WorkspaceItem.elm:870]: This didn't work with actual projects. I'll check and fix it soon

**Comment by yoching** [comment]
I have updated the codebase, now it works to some extent. However, I found a bug.

After opening `PositiveInt2`, when you click the link to type `PositiveInt2`in the term declaration, it tries to fetch the definition infinitely somehow. I tried to debug it, but quite difficult. Were there similar issues before?

Screenshot 2024-06-21 at 16 33 18

**Comment by hojberg** [comment]
@yoching thanks for taking this on! I haven't seen that infinite loop bug before. Any thoughts on how we might attempt to verify it on Share (I'm not sure where `PositiveInt2` is from?)

**Comment by hojberg** [comment]
I found it! It doesn't seem to loop for me: https://share.unison-lang.org/@kristiannotari/psql/code/main/latest/terms/examples/API/types/PositiveInt2

**Comment by yoching** [comment]
@hojberg 
The infinite fetch happens after this change. I'm testing local-ui with ui-core on this branch.

**Comment by yoching** [comment]
https://github.com/yoching/ui-core/blob/fix-multiple-definition/src/Code/Workspace.elm#L410-L412

**Comment by yoching** [comment]
It seems to make sense to make references based on the API response for items in `workspaceItems`, as API might response with multiple items for one request (not handling this was the source of the original issue #68).

**Comment by yoching** [comment]
One of the potential issues might be here in local-ui:
https://github.com/unisonweb/unison-local-ui/blob/main/src/UnisonLocal/Page/CodePage.elm#L131-L137

Regardless of the Msg sent, it's calling `updateSubpage`, leading to the call of `Workspace.open`. It might be better to change the logic based on Msg received.

**Comment by yoching** [comment]
This is where reference is constructed based on API response
https://github.com/unisonweb/ui-core/pull/90#discussion_r1621271414

**Comment by yoching** [comment]
Working on this issue, might take a bit more

**Comment by hojberg** [comment]
Ah yeah I've dealt with loops from `updateSubpage`, its not fun!

**Comment by yoching** [comment]
Fixed loop fetch issue with a different approach!

**Comment by yoching** [comment]
@hojberg 
It seems UI.Switch is not used anymore, which triggers the Elm review error on CI. Is it ok to remove it from codebase?

**Comment by hojberg** [comment]
@yoching yeah I think thats good to remove. Sorry for sitting on this PR for so long. Lets get this passing and merged

**Comment by yoching** [comment]
Hello. I've removed UI.Switch. I think it's ready for review.
I also have an open PR here: https://github.com/unisonweb/ui-core-scripts/pull/2, but it seems local-ui doesn't work with latest ui-core anymore?

**Review by hojberg — APPROVED** [review]

**Comment by hojberg** [comment]
Thank you @yoching !! Sorry for taking so long to get this merged.
local-ui is meant to continue to work with ui-core, though it will soon be replaced by https://github.com/unisonweb/ucm-desktop

**Pull request merged**